### DISCUSSION TITLE
Замечания по Ru Core IG

### DISCUSSION THREAD

**Discussion by eugenekogan-spb** [fhir-ru/core#266] [General]
Разбираем замечания к рабочей версии RuCoreIG

**Comment by eugenekogan-spb** [comment]
1. Build warnings
<img width="1223" alt="image" src="https://github.com/user-attachments/assets/69a47f7f-c976-4a1d-9b4d-6e7cfd4834e2">

**Reply by Vrodvi** [reply]
Нашёл вот такое обсуждение - https://chat.fhir.org/#narrow/stream/196008-ig-publishing-requirements/topic/Implementation.20Guide.20Page. 


Грэхем и компания обсудили, что есть 4 фрагмента метаданных про руководство, которые не включены в стандарный шаблон сайта руководства, но которые важно отображать. 

Вот эти 4 фрагмента:

- ip-statements.xhtml - сводка по интеллектуальной собственности, используемой в руководстве
![image](https://github.com/user-attachments/assets/8d946b76-e63c-4171-aa2c-27aa5489cb87)


- cross-version-analysis.xhtml - информация по использованной версии FHIR
![image](https://github.com/user-attachments/assets/214df885-ae4d-40e9-b71a-752e79b71ef3)


- dependency-table.xhtml - сводка по внешним связям руководства
![image](https://github.com/user-attachments/assets/99efc902-06f8-45a3-82fe-37f749c1fdc1)


- globals-table.xhtml - сводка по определённым в руководстве глобальным расширениям
![image](https://github.com/user-attachments/assets/2b724c52-002a-49e9-bef1-a4ca90f53b27)


---

Поскольку эти фрагменты очень важны для интерпретации руководства, то была добавлена проверка в сборщик сайта руководства, чтобы проверялось наличие этих важных фрагментов. 

Авторы руководства самостоятельно решают куда вставить эти фрагменты, но самый простой вариант - это размещение прямо на заглавной странице.
Вставить данные фрагменты в руководство можно следующим образом - всатвить {% include ТУТ_НАЗВАНИЕ_ФРАГМЕНТА.xhtml %} в любой из текстовых маркдаун-файлов руководства (например в index.md), например:
```
# RuCoreIG

Feel free to modify this index page with your own awesome content!

# IG metadata

{% include ip-statements.xhtml %}

{% include cross-version-analysis.xhtml %}

{% include dependency-table.xhtml %}

{% include globals-table.xhtml %}

```

После добавления этих фрагментов предупреждение уходит, а результирующее руководство дополняется данными этих фргаментов:
![image](https://github.com/user-attachments/assets/1f7e052a-9aad-41d1-bf38-5f993f397dbe)

**Reply by eugenekogan-spb** [reply]
Есть ли возможность отредактировать содержание ip-statements? Наше руководство не содержит материалы LOINC. Кроме того, указание об использовании HL7 Terminology публикуется дважды.

**Comment by eugenekogan-spb** [comment]
2. Name should be usable as an identifier 
<img width="1396" alt="image" src="https://github.com/user-attachments/assets/41703a50-36aa-43ba-a0ad-aa25b28ddda8">

**Reply by Vrodvi** [reply]
Проблема в атрибуте с машиночитаемым именем артефакта - `* name =   "core-ns-nsi-confidentiality"`.

Правильный вариант -  `*   name =   "Core_Ns_Nsi_Confidentiality"` - с заглавных букв и с нижним подчёркиванием в качестве разделителя:
```
Instance:   core-ns-nsi-confidentiality 
InstanceOf: NamingSystem
Title: "Core NamingSystem NSI confidentiality (Уровень конфиденциальности медицинского документа)"
Usage: #definition

//* name =   "core-ns-nsi-confidentiality"
*   name =   "Core_Ns_Nsi_Confidentiality"
```

![image](https://github.com/user-attachments/assets/b7c5ac30-c0ea-42b8-89dc-7e6d1dfa302e)

**Reply by eugenekogan-spb** [reply]
Fixed

**Comment by eugenekogan-spb** [comment]
3. Organisation Instance: Identifier [INN].type.coding has no system
<img width="1302" alt="image" src="https://github.com/user-attachments/assets/f6ada53e-3507-4d7a-b3c4-2c9d4967ff19">

**Reply by Vrodvi** [reply]
В атрибуте `type` в профиле и в экземпляре указн один только код без указания системы кодирования, из которой взят этот код.
Поскольку тип данных у атрибута `type` - не `code` (для которого допустимо указание кода без системы), а `CodeableConcept`, то кодирование обязательно должно осуществляться с указанием системы кодирования (в данном случае код  #TAX взят из стандартной системы http://terminology.hl7.org/CodeSystem/v2-0203 - её и нужно указать).

Т.е. вместо:
```
Profile: Core_Organization

  * type = #TAX

Instance: core-organization-instance-full

  * type = #TAX

```

должно быть:
```
Profile: Core_Organization

  * type = http://terminology.hl7.org/CodeSystem/v2-0203#TAX

Instance: core-organization-instance-full

  * type = http://terminology.hl7.org/CodeSystem/v2-0203#TAX
```

![image](https://github.com/user-attachments/assets/b82beb65-0fd0-42c6-a837-ef01aed0b1ef)

**Reply by eugenekogan-spb** [reply]
@Vrodvi 

Когда я для CoreOrganization указываю фиксированное значение:
  * type = http://terminology.hl7.org/CodeSystem/v2-0203#TAX

сборщик даёт следующую ошибку (алиасы убраны, лог пишется корректный):

Sushi: error Cannot assign #TAX to this element; a different CodeableConcept is already assigned: {"coding":[{"code":"TAX","system":"http://terminology.hl7.org/CodeSystem/v2-0203"}]}. (00:00.187 / 00:18.145, 97Mb)
Sushi:   File: /scratch/repo/input/fsh/CoreOrganization.fsh                                          (00:00.000 / 00:18.145, 97Mb)
Sushi:   Line: 83                                                                                    (00:00.000 / 00:18.146, 97Mb)
Sushi: error Cannot assign #TAX to this element; a different CodeableConcept is already assigned: {"coding":[{"code":"TAX","system":"http://terminology.hl7.org/CodeSystem/v2-0203"}]}. (00:00.090 / 00:18.236, 97Mb)
Sushi:   File: /scratch/repo/input/fsh/CoreOrganizationPayer.fsh                                     (00:00.000 / 00:18.236, 97Mb)
Sushi:   Line: 36  

Поскольку ругается и на CoreOrganization и на CoreOrganizationPayer, в котором нет своих определений, то возможно это сообщение относится к экземплярам?

**Reply by eugenekogan-spb** [reply]
Да, вопрос снят. Сообщение об ошибке относилось именно к указанию значения атрибута identifier.type в экземплярах ресурсов, в ситуации, когда профилем определено фиксированное значение этого атрибута.

**Reply by eugenekogan-spb** [reply]
Fixed

**Comment by eugenekogan-spb** [comment]
4. Patient Instance: a definition for code system could not be found
<img width="1390" alt="image" src="https://github.com/user-attachments/assets/7c9107f5-85c8-4133-9ed1-114bc532acaf">

Почему-то замечания касаются только этой системы кодирования.

**Reply by Vrodvi** [reply]
Это наша локальная система кодирования и у неё указан тип наполнения `* ^content = #example`, т.е. наполнение не полное, а только пример.
При этом на данной системе кодирования составлен набор значений и он использован в профиле пациента.

Публикатор не может оперировать примерным наполнением системы кодирования, поэтому пытается найти полное наполнение этой системы в стандартном публичном терминологическом сервисе `http://tx.fhir.org/r4`, где система отсутствует.
Следовательно и провалидировать экземпляр ресурса по профилю с таким набором значений публикатор не может - о чём выводится предупреждение.

При установке типа наполнения равной `* ^content = #complete` публикатор использует только указанное наполнение и никуда во вне не ходит, что исправляет ошибку.
```
CodeSystem: Core_Cs_Nsi_Identity_Document

* ^content = #complete
```

![image](https://github.com/user-attachments/assets/e328974a-efa6-4611-9249-9ab25f1532d7)

**Reply by eugenekogan-spb** [reply]
Поставил complete для всех систем кодирования, кроме:
- должности медицинских работников
- регистр медицинских организаций
- типы медицинских документов

для которых указал not-present.

Предупреждения остались только для этих систем. Предлагаю пока оставить.

**Comment by eugenekogan-spb** [comment]
5. Review the extension type 
<img width="1385" alt="image" src="https://github.com/user-attachments/assets/9e508c48-3263-4644-8532-22ee84e42c35">

**Reply by Vrodvi** [reply]
В FSH предусмотрена возможность уточнить контекст, в рамках которого возможно использование созданного расширения - см. https://build.fhir.org/ig/HL7/fhir-shorthand/reference.html#defining-extensions

В нашем случае контекст действия расширения - это тип данных [Address](https://hl7.org/fhir/datatypes.html#Address).
Если обозначить это явно через атрибут `Context:`, то предупреждение пропадает:
![image](https://github.com/user-attachments/assets/b508778e-d21c-4185-86fe-3615988ad93b)

Т.е. вместо:
```
Extension: Core_Ex_Address_Fias
Id:        core-ex-address-fias
Title: "Core Extension address fias (Кодирование адреса по ФИАС)"
Description: "Кодирование адреса по ФИАС. При маппинге на СЭМД данный url соответствует urn:hl7-ru:fias"
```
необходимо сделать:
```
Extension: Core_Ex_Address_Fias
Id:        core-ex-address-fias
Title: "Core Extension address fias (Кодирование адреса по ФИАС)"
Description: "Кодирование адреса по ФИАС. При маппинге на СЭМД данный url соответствует urn:hl7-ru:fias"
Context: Address
```

**Reply by eugenekogan-spb** [reply]
Fixed.

**Comment by eugenekogan-spb** [comment]
6. Description мы написали, но на сайте его нет.
<img width="421" alt="image" src="https://github.com/user-attachments/assets/0fb85f15-e393-47c3-a465-f1f2bb2e49d7">
<img width="724" alt="image" src="https://github.com/user-attachments/assets/e04edb6c-ed2a-43a6-9aab-1457efb79f06">

**Reply by Vrodvi** [reply]
На примере mCode https://build.fhir.org/ig/HL7/fhir-mCODE-ig/Patient-cancer-patient-adam-everyman.html такая же ситуация - description не выводится, плюс они Title даже не указывают и на страничке выводится идентификатор экземпляра.

Но в списке артефактов у них, всё же, адекватное название - они формируют его вручную через markdown-страничку с ручным перечнем всех примеров и человекочитаемым описанием каждого примера:
![image](https://github.com/user-attachments/assets/c2ffec6f-321b-4def-ac1a-0255b8d96307)

**Reply by Vrodvi** [reply]
Перепроверил. 

Для экземпляров:
- title отображается в списке артефактов и в карточке экземпляра
- description нигде не выводится
- если не заполнить title и description, то для отображения будет использован идентификатор (из `Instance:`)

Для профилей:
- title отображается в списке артефактов и в карточке экземпляра
- description отображается в списке артефактов и в карточке экземпляра
- в карточке профиля отображается id связанного с ним экземпляра, а не его title или description 
- если не заполнить title и description, то для отображения будет использовано машинное имя (из `Profile:`)

![image](https://github.com/user-attachments/assets/2f847a0c-1ddc-42f1-9580-ca1b7e32da03)


---

для проверки использованы следующие тестовые артефакты:
```
// ----------------------------------------------------------------- ТЕСТ 1 ---------------------------------------------------------------- \\
// проверяем вывод ID, TITLE, DESCRIPTION
// вариант: заполнены все атрибуты - ID, TITLE, DESCRIPTION

Profile: Core_ServiceRequest_ForTEST1
Id:      core-servicerequest-fortest1
Parent:       ServiceRequest
Title: "Core Service Request (Заказ услуги) ТЕСТ1: ПРОВЕРЯЕМ ВЫВОД - Я ИЗ TITLE, ЗАПОЛНЕНЫ ID, TITLE И DESCRIPTION"
Description: "Заказ услуги ТЕСТ1: ПРОВЕРЯЕМ ВЫВОД - Я ИЗ DESCRIPTION, ЗАПОЛНЕНЫ ID, TITLE И DESCRIPTION"
* subject 1..1
* subject ^short = "Пациент"
* subject only Reference(Core_Patient)

Instance:   core-servicerequest-fortest1-instance-full
InstanceOf: Core_ServiceRequest_ForTEST1
Title: "Core Instance Service Request Full ТЕСТ1: ПРОВЕРЯЕМ ВЫВОД - Я ИЗ TITLE, ЗАПОЛНЕНЫ  ID, TITLE И DESCRIPTION"
Description: "запрос услуги (неполный! - непонятно что заказано) ТЕСТ1: ПРОВЕРЯЕМ ВЫВОД - Я ИЗ DESCRIPTION, ЗАПОЛНЕНЫ  ID, TITLE И DESCRIPTION"
Usage: #example
* status = #active
* intent = #plan
* subject = Reference(core-patient-instance-full)

// ----------------------------------------------------------------- ТЕСТ 2 ---------------------------------------------------------------- \\
// проверяем вывод ID, TITLE, DESCRIPTION
// вариант: заполнены только ID и TITLE

Profile: Core_ServiceRequest_ForTEST2
Id:      core-servicerequest-fortest2
Parent:       ServiceRequest
Title: "Core Service Request (Заказ услуги) ТЕСТ2: ПРОВЕРЯЕМ ВЫВОД - Я ИЗ TITLE, ЗАПОЛНЕН ТОЛЬКО ID и TITLE"
* subject 1..1
* subject ^short = "Пациент"
* subject only Reference(Core_Patient)

Instance:   core-servicerequest-fortest2-instance-full
InstanceOf: Core_ServiceRequest_ForTEST2
Title: "Core Instance Service Request Full ТЕСТ2: ПРОВЕРЯЕМ ВЫВОД - Я ИЗ TITLE, ЗАПОЛНЕН ТОЛЬКО ID и TITLE"
Usage: #example
* status = #active
* intent = #plan
* subject = Reference(core-patient-instance-full)

// ----------------------------------------------------------------- ТЕСТ 3 ---------------------------------------------------------------- \\
// проверяем вывод ID, TITLE, DESCRIPTION
// вариант: заполнены только ID и DESCRIPTION

Profile: Core_ServiceRequest_ForTEST3
Id:      core-servicerequest-fortest3
Parent:       ServiceRequest
Description: "Заказ услуги ТЕСТ3: ПРОВЕРЯЕМ ВЫВОД - Я ИЗ DESCRIPTION, ЗАПОЛНЕН ТОЛЬКО ID и DESCRIPTION"
* subject 1..1
* subject ^short = "Пациент"
* subject only Reference(Core_Patient)

Instance:   core-servicerequest-fortest3-instance-full
InstanceOf: Core_ServiceRequest_ForTEST3
Description: "запрос услуги (неполный! - непонятно что заказано) ТЕСТ3: ПРОВЕРЯЕМ ВЫВОД - Я ИЗ DESCRIPTION, ЗАПОЛНЕН ТОЛЬКО ID и DESCRIPTION"
Usage: #example
* status = #active
* intent = #plan
* subject = Reference(core-patient-instance-full)

// ----------------------------------------------------------------- ТЕСТ 4 ---------------------------------------------------------------- \\
// проверяем вывод ID, TITLE, DESCRIPTION
// вариант: заполнен только ID

Profile: Core_ServiceRequest_ForTEST4_Only_Id
Id:      core-servicerequest-fortest4-only-id
Parent:       ServiceRequest
* subject 1..1
* subject ^short = "Пациент"
* subject only Reference(Core_Patient)

Instance:   core-servicerequest-fortest4-only-id-instance-full
InstanceOf: Core_ServiceRequest_ForTEST4_Only_Id
Usage: #example
* status = #active
* intent = #plan
* subject = Reference(core-patient-instance-full)
```

**Comment by Vrodvi** [comment]
Ещё одно замечание публикатора, которое можно легко поправить:
![image](https://github.com/user-attachments/assets/82674b03-79e0-40d1-a492-a2e9cb0d014e)
- тут для справочника, содержащего иерархически вложенные коды явно не обозначен смысл этой иерархии. Рекомендуется явно прописыть что обозначает иерархия. Обозначать необходимо по справочнику https://www.hl7.org/fhir/valueset-codesystem-hierarchy-meaning.html

В данной ситуации смысл иерархии - `is-a`, что можно обозначить в профиле следующим образом:
```
CodeSystem: Core_Cs_Nsi_Identity_Document
...
* ^hierarchyMeaning = #is-a
...
```
что устраняет информационное сообщение:

![image](https://github.com/user-attachments/assets/fee2f961-67b4-4331-ad2b-f156af550d42)
